addons/stock/models/stock_move_line.py

```diff
@@ -451,12 +452,61 @@ def _action_done(self, cancel_backorder=False):
                         Quant._update_available_quantity(ml.product_id, ml.location_id, taken_from_untracked_qty, lot_id=ml.lot_id, package_id=ml.package_id, owner_id=ml.owner_id)
                 Quant._update_available_quantity(ml.product_id, ml.location_dest_id, quantity, lot_id=ml.lot_id, package_id=ml.result_package_id, owner_id=ml.owner_id, in_date=in_date)
             done_ml |= ml
+
+            ml._check_duplicates()
+
         # Reset the reserved quantity as we just moved it to the destination location.
         (self - ml_to_delete).with_context(bypass_reservation_update=True).write({
             'product_uom_qty': 0.00,
             'date': fields.Datetime.now(),
         })
 
+    def _check_duplicates(self):
+        # Unicity of serial tracked product. Search in quants if the sn is
+        # already present in stock or delivered to a customer
+        if self.tracking == 'serial' and self.lot_id:
+            similar_quant = self.env['stock.quant'].search([
+                ('lot_id', '=', self.lot_id.id),
+                '|',
+                ('location_id.usage', 'not in', ('supplier', 'inventory', 'production')),
+                ('location_id.scrap_location', '=', True)
+            ])
+            error_message = _('The serial number %s for product %s is already in another location. Correct your inventory with an inventory adjustment before validating this product move.' % (self.lot_id.name, self.product_id.display_name))
+
+            # Check if SN has been scrapped
+            scrapped = any([quant.location_id.scrap_location for quant in similar_quant])
+            if scrapped and not self.location_dest_id.scrap_location:
```

Comment:
not not

Reply:
si la move line en question mène à Scrap, alors le move est légitime. Sinon on raise